### PR TITLE
feat: Add CUDA support

### PR DESCRIPTION
**Pull request by vmx**
A new environment variable `BELLMAN_GPU_FRAMEWORK` is introduced to select
CUDA or OpenCL if Bellperson is compiled with support for both.

Currently only the multiexp calculation is supported by CUDA, the FFT
calculation still runs on OpenCL.

BREAKING CHANGE: `gpu` feature is replaced with `cuda` and `opencl` feature.

The `gpu` feature is now split into two pieces, `opencl` and `cuda`, you can
compile the library with either or both.

### CODE REVIEW

**Comment by vmx** [comment]
Ready for review. The CI is green and `rust-gpu-tools` has been updated to point to the master branch, so this PR is ready to be merged after being reviewed.

File: src/gpu/program.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+#[cfg(not(all(feature = "cuda", feature = "opencl")))]
+use crate::gpu::error::GPUError;
+use crate::gpu::error::GPUResult;
+#[cfg(feature = "opencl")]
+use crate::gpu::sources;
+use crate::gpu::GpuEngine;
+use pairing::Engine;
+
+/// Returns the program for the preferred [`rust_gpu_tools::device::Framework`].
```
Comment (dignifiedquire) [inline src/gpu/program.rs:20]: this is very similar than the neptune version, should this be put into rust-gpu-tools in an abstract fashion?

File: src/gpu/program.rs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+///
+/// If the device supports CUDA, then CUDA is used, else OpenCL. You can force a selection with
+/// the environment variable `BELLMAN_GPU_FRAMEWORK`, which can be set either to `cuda` or `opencl`.
+pub fn program<E>(device: &Device) -> GPUResult<Program>
+where
+    E: Engine + GpuEngine,
+{
+    let framework = match env::var("BELLMAN_GPU_FRAMEWORK") {
```
Comment (dignifiedquire) [inline src/gpu/program.rs:28]: can you add a test for this? I have added a helper in `multicore` for testing env variable based behaviour

File: src/gpu/program.rs (hunk 1 above)
Comment (vmx) [reply src/gpu/program.rs:20]: I had the same thought. The problem is that the part below (https://github.com/filecoin-project/bellperson/pull/218/files/4a577d76bb62c0338c54c408a3ee8eea17b4e1fb#diff-48d758a332b8b000b1386054877cf659fcc40ae7a2e4163552b0ad4659d2f1a9R60-R77) is slightly different (and will always differ between libraries (also the returned errors are different). Hence I'd keep that duplicated piece of code for now.

File: src/gpu/program.rs (hunk 2 above)
Comment (vmx) [reply src/gpu/program.rs]: I've moved the helper into its own module.

**Review by dignifiedquire — APPROVED** [review]

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**